spec/app/ics-721-nft-transfer/README.md

```diff
@@ -0,0 +1,385 @@
+    if source {
+      // receiver is source chain: unescrow token
+      // determine escrow account
+      escrowAccount = channelEscrowAddresses[packet.destChannel]
+      // unescrow token to receiver
+      err = nft.Transfer(data.classId.slice(len(prefix)), data.tokenIds[i], data.receiver)
+      if (err !== nil) {
+        ack = NonFungibleTokenPacketAcknowledgement{false, "transfer nft(" + data.classId + ", " + data.tokenIds[i] + ") failed"}
+        break
+      }
+    } else {
+      prefix = "{packet.destPort}/{packet.destChannel}/"
+      prefixedClassId = prefix + data.classId
+      // sender was source, mint voucher to receiver
+      err = nft.Mint(prefixedClassId, data.classUri, data.tokenIds[i], data.tokenUris[i], data.receiver)
+      if (err !== nil) {
+        ack = NonFungibleTokenPacketAcknowledgement{false, "mint nft(" + data.classId + ", " + data.tokenIds[i] + ") failed"}
+        break
+      }
+    }
+  }
+  return ack
+}
+```
+
+`onAcknowledgePacket` is called by the routing module when a packet sent by this module has been acknowledged.
+
+```typescript
+function onAcknowledgePacket(
+  packet: Packet,
+  acknowledgement: bytes) {
+  // if the transfer failed, refund the tokens
+  if (!ack.success)
+    refundToken(packet)
+}
+```
+
+`onTimeoutPacket` is called by the routing module when a packet sent by this module has timed out (such that it will not be received on the destination chain).
+
+```typescript
+function onTimeoutPacket(packet: Packet) {
+  // the packet timed-out, so refund the tokens
+  refundToken(packet)
+}
+```
+
+`refundToken` is called by both `onAcknowledgePacket`, on failure, and `onTimeoutPacket`, to refund escrowed token to the original sender.
+
+```typescript
+function refundToken(packet: Packet) {
+  NonFungibleTokenPacketData data = packet.data
+  prefix = "{packet.sourcePort}/{packet.sourceChannel}/"
+  for (let tokenId in data.tokenIds) {
+    // we are the source if the classId is not prefixed with the packet's sourcePort and sourceChannel
+    source = data.classId.slice(0, len(prefix)) !== prefix
+    if source {
+      // sender was source chain, unescrow tokens back to sender
+      escrowAccount = channelEscrowAddresses[packet.srcChannel]
+      // assert that escrow account is token owner
+      abortTransactionUnless(escrowAccount === nft.getOwner(data.classId, tokenId))
+      nft.Transfer(data.classId, tokenId, data.sender)
+    } else {
+      // receiver was source chain, mint voucher back to sender
+      nft.Mint(data.classId, tokenId, data.sender)
+    }
+  }
+}
+```
+
+```typescript
+function onTimeoutPacketClose(packet: Packet) {
+  // can't happen, only unordered channels allowed
+}
+```
+
+#### Reasoning
+
+##### Correctness
+
+This implementation preserves token uniqueness.
+
+Uniqueness: If tokens have been sent to the counterparty chain, they can be redeemed back in the same `classId` & `tokenId` on the source chain.
+
+##### Multi-chain notes
```

Comment:
Maybe refer here to ICS20 instead. It's exactly the same content (except `classId` instead of denomination).

Reply:
ICS20 talks about fungibility and supply here, which are two concepts that really don't apply to NFTs, as Chris commented earlier.

Reply:
Will update the Multi-chain notes section to make it more clear